In [1]:
from nltk.corpus import wordnet

<b>1) Найти все значения (синсеты) для лексемы nail<b>

In [19]:
nail = wordnet.synsets('nail')
for ss in nail:
    print(ss, ss.definition())

Synset('nail.n.01') horny plate covering and protecting part of the dorsal surface of the digits
Synset('nail.n.02') a thin pointed piece of metal that is hammered into materials as a fastener
Synset('nail.n.03') a former unit of length for cloth equal to 1/16 of a yard
Synset('nail.v.01') attach something somewhere by means of nails
Synset('collar.v.01') take into custody
Synset('smash.v.01') hit hard
Synset('nail_down.v.01') succeed in obtaining a position
Synset('breeze_through.v.01') succeed at easily
Synset('pinpoint.v.01') locate exactly
Synset('complete.v.04') complete a pass


<b>2) Найти определение для лексемы nail в значении (а) "ноготь" и в значении (b) "гвоздь"</b>

In [20]:
print(nail[0], nail[0].definition())
print(nail[1], nail[1].definition())

Synset('nail.n.01') horny plate covering and protecting part of the dorsal surface of the digits
Synset('nail.n.02') a thin pointed piece of metal that is hammered into materials as a fastener


<b>3) Найдите два произвольных контекста для слова nail в значениях (a) "ноготь" и (b) "гвоздь"; продемонстрируйте на них действие алгоритма Леска для разрешения неоднозначности</b>

In [21]:
sent1 = 'She went to beauty salon to paint her nails.'
sent2 = 'She hangs a new painting with a nail.'
sent1_tokens = [word.strip('.,') for word in sent1.split(' ')]
sent2_tokens = [word.strip('.,') for word in sent2.split(' ')]
print(sent1_tokens)
print(sent2_tokens)

['She', 'went', 'to', 'beauty', 'salon', 'to', 'paint', 'her', 'nails']
['She', 'hangs', 'a', 'new', 'painting', 'with', 'a', 'nail']


In [22]:
from nltk.wsd import lesk
print(lesk(sent1_tokens, 'nail').definition())
print(lesk(sent2_tokens, 'nail').definition())

attach something somewhere by means of nails
succeed in obtaining a position


Алгоритм Леска ошибся в обоих случаях. Более того, он ошибся даже с частью речи. Попробуем ограничить часть речи существительными.

In [23]:
print(lesk(sent1_tokens, 'nail', 'n').definition())
print(lesk(sent2_tokens, 'nail', 'n').definition())

a former unit of length for cloth equal to 1/16 of a yard
a former unit of length for cloth equal to 1/16 of a yard


Не помогло :(

<b>4) Найдите гиперонимы для значения (a) и гиперонимы для значения (b)</b>

In [26]:
for ss in nail[0].hypernyms():
    print(ss, ss.definition())
for ss in nail[1].hypernyms():
    print(ss, ss.definition())

Synset('horny_structure.n.01') any rigid body structure composed primarily of keratin
Synset('fastener.n.02') restraint that attaches to something or holds something in place


<b>5) Вычислите наименьшее расстояние между значением nail "ноготь" и значениями лексемы finger, а также nail "гвоздь" и значениями лексемы hammer.
Найти min (d(nail: "ноготь", finger), d(nail: "ноготь", hammer)), а также min (d(nail: "гвоздь", finger), d(nail: "гвоздь", hammer))</b>

In [27]:
finger = wordnet.synsets('finger')
hammer = wordnet.synsets('hammer')

def get_dist_sim(ss1, lexeme):
    distances = []
    similarities = []
    for ss2 in lexeme:
        dist = ss1.shortest_path_distance(ss2)
        if dist is not None:
            distances.append(dist)
            sim = ss1.path_similarity(ss2)
            similarities.append(sim)
    return distances, similarities

# min d(plant: "ноготь", finger)
dist1 = get_dist_sim(plant[0], finger)[0]
print('min d(plant: "ноготь", finger): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(finger[dist1.index(min(dist1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# min d(plant: "ноготь", hammer)
dist2 = get_dist_sim(plant[0], hammer)[0]
print('min d(plant: "ноготь", hammer): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(hammer[dist2.index(min(dist2))].definition()))

# min d(plant: "гвоздь", finger)
dist3 = get_dist_sim(plant[1], finger)[0]
print('min d(plant: "гвоздь", finger): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(finger[dist3.index(min(dist3))].definition()))

# min d(plant: "гвоздь", hammer)
dist4 = get_dist_sim(plant[1], hammer)[0]
print('min d(plant: "гвоздь", hammer): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(hammer[dist4.index(min(dist4))].definition()))

print('min (d(plant: "ноготь", finger), d(plant: "ноготь", hammer)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(plant: "гвоздь", finger), d(plant: "гвоздь", hammer)): {}'.format(min(min(dist3), min(dist4))))

min d(plant: "ноготь", finger): 6
closest lemma definition: any of the terminal members of the hand (sometimes excepting the thumb)

min d(plant: "ноготь", hammer): 10
closest lemma definition: the ossicle attached to the eardrum

min d(plant: "гвоздь", finger): 7
closest lemma definition: one of the parts of a glove that provides covering for a finger or thumb

min d(plant: "гвоздь", hammer): 6
closest lemma definition: a power tool for drilling rocks

min (d(plant: "ноготь", finger), d(plant: "ноготь", hammer)): 6
min (d(plant: "гвоздь", finger), d(plant: "гвоздь", hammer)): 6


Кратчайшее расстояние от ногтя до finger - 6, до hammer - 10. Получается, что до finger расстояние короче, что логично.
Кратчайшее расстояние от гвоздя до finger - 7, до hammer - 6. До hammer расстояние короче, что ожидаемо. Неожиданно — близкое расстояние от гвоздя до finger.

Если имелось в виду посчитать не расстояние, а сходство путей на основе кратчайшего расстояния (минимальный путь --> максимальное сходство), вот оно:

In [28]:
# max s(nail: "ноготь", finger)
sim1 = get_dist_sim(nail[0], finger)[1]
print('max s(nail: "ноготь", finger): {}'.format(max(sim1)))
print('closest lemma definition: {}\n'.format(finger[sim1.index(max(sim1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# max s(nail: "ноготь", hammer)
sim2 = get_dist_sim(nail[0], hammer)[1]
print('max s(nail: "ноготь", hammer): {}'.format(max(sim2)))
print('closest lemma definition: {}\n'.format(hammer[sim2.index(max(sim2))].definition()))

# max s(nail: "гвоздь", finger)
sim3 = get_dist_sim(nail[1], finger)[1]
print('max s(nail: "гвоздь", finger): {}'.format(min(sim3)))
print('closest lemma definition: {}\n'.format(finger[sim3.index(max(sim3))].definition()))

# max s(nail: "гвоздь", hammer)
sim4 = get_dist_sim(nail[1], hammer)[1]
print('max s(nail: "гвоздь", hammer): {}'.format(max(sim4)))
print('closest lemma definition: {}\n'.format(hammer[sim4.index(max(sim4))].definition()))

print('max (s(nail: "ноготь", finger), s(nail: "ноготь", hammer)): {}'.format(max(max(sim1), max(sim2))))
print('max (s(nail: "гвоздь", finger), s(nail: "гвоздь", hammer)): {}'.format(max(max(sim3), max(sim4))))

max s(nail: "ноготь", finger): 0.14285714285714285
closest lemma definition: any of the terminal members of the hand (sometimes excepting the thumb)

max s(nail: "ноготь", hammer): 0.09090909090909091
closest lemma definition: the ossicle attached to the eardrum

max s(nail: "гвоздь", finger): 0.0625
closest lemma definition: one of the parts of a glove that provides covering for a finger or thumb

max s(nail: "гвоздь", hammer): 0.14285714285714285
closest lemma definition: a power tool for drilling rocks

max (s(nail: "ноготь", finger), s(nail: "ноготь", hammer)): 0.14285714285714285
max (s(nail: "гвоздь", finger), s(nail: "гвоздь", hammer)): 0.14285714285714285


Результат тот же (что ожидаемо, потому что используется фактически одна и та же мера). 

<b>6) Вычислить двумя разными способами расстояние:
d(nail: "гвоздь", Galvanized) и d(organism, whole)
Есть ли разница в расстояниях? Какое из расстояний, по Вашему мнению, в лучшей степени отражает интуитивное представление о семантчиеской близости слов?</b>

А вот здесь же точно речь идет о сходстве/близости значений, а не о стандартном расстоянии? (Потому что стандартное расстояние - кратчайший путь - трудно считать несколькими способами.)

In [36]:
helix = wordnet.synsets("helix")
for ss in master:
    print(ss, ss.definition())

Synset('helix.n.01') a curve that lies on the surface of a cylinder or cone and cuts the element at a constant angle
Synset('coil.n.01') a structure consisting of something wound in a continuous series of loops
Synset('helix.n.03') type genus of the family Helicidae


Посчитаем близость с помощью критериев Path Similarity, Leacock-Chodorow Similarity, Wu-Palmer Similarity, Resnik Similarity, Jiang-Conrath Similarity и Lin Similarity. Для последних трех будем использовать Information Content корпуса Brown.

In [37]:
from nltk.corpus import wordnet_ic
ic = wordnet_ic.ic('ic-brown.dat')

print(nail[1].path_similarity(helix[0]))

#Leacock-Chodorow Similarity
print(nail[1].lch_similarity(helix[0]))

#Wu-Palmer Similarity
print(nail[1].wup_similarity(helix[0]))
    
#Resnik Similarity
print(nail[1].res_similarity(helix[0], ic))
    
#Jiang-Conrath Similarity
print(nail[1].jcn_similarity(helix[0], ic))
    
#Lin Similarity
print(nail[1].lin_similarity(helix[0], ic))


0.0625
0.8649974374866046
0.11764705882352941
-0.0
0.042816924205233774
-0.0


In [38]:
organism = wordnet.synsets('organism')
for ss in organism:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('organism.n.01') a living thing that has (or can develop) the ability to act or function independently
Synset('organism.n.01') []
Synset('organism.n.02') a system considered analogous in structure or function to a living body
Synset('organism.n.02') ['the social organism']


In [39]:
whole = wordnet.synsets('whole', 'n')
for ss in whole:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('whole.n.01') all of something including all its component elements or parts
Synset('whole.n.01') ['Europe considered as a whole', 'the whole of American literature']
Synset('whole.n.02') an assemblage of parts that is regarded as a single entity
Synset('whole.n.02') ['how big is that part compared to the whole?', 'the team is a unit']


In [40]:
for ss1 in organism:
    print(ss1.definition())
    for ss2 in whole:
        print(ss2.definition())
        print(ss1.path_similarity(ss2))

        #Leacock-Chodorow Similarity
        print(ss1.lch_similarity(ss2))

        #Wu-Palmer Similarity
        print(ss1.wup_similarity(ss2))

        #Resnik Similarity
        print(ss1.res_similarity(ss2, ic))

        #Jiang-Conrath Similarity
        print(ss1.jcn_similarity(ss2, ic))

        #Lin Similarity
        print(ss1.lin_similarity(ss2, ic))

a living thing that has (or can develop) the ability to act or function independently
all of something including all its component elements or parts
0.07692307692307693
1.072636802264849
0.14285714285714285
-0.0
0.11927453545173074
-0.0
an assemblage of parts that is regarded as a single entity
0.3333333333333333
2.538973871058276
0.8
1.5318337432196856
1.4444255924181877
0.8156763478002892
a system considered analogous in structure or function to a living body
all of something including all its component elements or parts
0.1
1.3350010667323402
0.3076923076923077
0.5962292078977726
0.05928363701403374
0.0660256991738309
an assemblage of parts that is regarded as a single entity
0.125
1.55814461804655
0.2222222222222222
-0.0
0.07444639262208605
-0.0


<style type="text/css">
	table.tableizer-table {
		font-size: 12px;
		border: 1px solid #CCC; 
		font-family: Arial, Helvetica, sans-serif;
	} 
	.tableizer-table td {
		padding: 4px;
		margin: 3px;
		border: 1px solid #CCC;
	}
	.tableizer-table th {
		background-color: #104E8B; 
		color: #FFF;
		font-weight: bold;
	}
</style>
<table class="tableizer-table">
<thead><tr class="tableizer-firstrow"><th>Similarity Index</th><th>гвоздь/helix</th><th>a living thing / all of something</th><th>a living thing / an assemblage</th><th>a system / all of something</th><th>a system / an assemblage</th></tr></thead><tbody>
 <tr><td>Path Similarity</td><td>0.0625</td><td>0.07692307692</td><td>0.3333333333</td><td>0.1</td><td>0.125</td></tr>
 <tr><td>Leacock-Chodorow Similarity</td><td>0.8649974374866046</td><td>1.072636802</td><td>2.538973871</td><td>1.335001067</td><td>1.558144618</td></tr>
 <tr><td>Wu-Palmer Similarity</td><td>0.11764705882352941</td><td>0.1428571429</td><td>0.8</td><td>0.3076923077</td><td>0.2222222222</td></tr>
 <tr><td>Resnik Similarity</td><td>0</td><td>0</td><td>1.531833743</td><td>0.5962292079</td><td>0</td></tr>
 <tr><td>Jiang-Conrath Similarity</td><td>0.042816924205233774</td><td>0.1192745355</td><td>1.444425592</td><td>0.05928363701</td><td>0.07444639262</td></tr>
 <tr><td>Lin Similarity</td><td>0</td><td>0</td><td>0.8156763478</td><td>0.06602569917</td><td>0</td></tr>
</tbody></table>


Helix просто является гипонимом гвоздя. Как его сравнивать с organism/whole не совсем понятно, потому что во втором случае нет таких прямых отношений. Leacock-Chodorow Similarity индекс показывает самый высокий показатель схожести.

По Resnik и Lin индексам схожесть между гвоздь/helix нулевая. Непонятно, говорит ли это о полной несхожести значений или это специфика индекса/корпуса/данных.

А вообще, значения индексов в основном довольно маленькие. Не совсем понятно, насколько целесообразно сравнение сотых и тысячных долей.

Если учитывать всё сразу, первые три индекса более или менее отражают интуитивное представление о близости слов.